src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include "software/ai/evaluation/calc_best_shot.h"
+#include "software/ai/evaluation/enemy_threat.h"
+#include "software/ai/hl/stp/tactic/move/move_fsm.h"
+#include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/geom/algorithms/distance.h"
+
+struct ShadowEnemyFSM
+{
+    class BlockState;
+    class StealAndChipState;
+
+    // this struct defines the unique control parameters that the ShadowEnemyFSM requires
+    // in its update
+    struct ControlParams
+    {
+        // The Enemy Threat indicating which enemy to shadow
+        std::optional<EnemyThreat> enemy_threat;
+
+        // How far from the enemy the robot will position itself to shadow
+        double shadow_distance;
+
+        // How close to the enemy before the robot will try to steal and chip the ball
+        double steal_and_chip_distance;
+    };
+
+    DEFINE_UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS
+
+
+    // Distance to chip the ball when trying to yeet it
+    // TODO (#1878): Replace this with a more intelligent chip distance system
+    static constexpr double YEET_CHIP_DISTANCE_METERS = 2.0;
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto block_s          = state<BlockState>;
+        const auto steal_and_chip_s = state<StealAndChipState>;
+
+        const auto update_e = event<Update>;
+
+        /**
+         * Guard that checks if the enemy threat has ball
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_has_ball = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return enemy_threat_opt.value().has_ball;
+            };
+            return false;
+        };
+
+        /**
+         * Guard that checks if the enemy threat is near
+         * our robot
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_is_near = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return distance(enemy_threat_opt.value().robot.position(),
+                                event.common.robot.position()) <=
+                       event.control_params.steal_and_chip_distance;
+            };
+            return false;
+        };
+
+        /**
+         * Action to block the ball
+         *
+         * If the enemy can pass to the robot we are shadowing, block the pass.
+         * Else, block the net
+         *
+         * @param event ShadowEnemyFSM::Update
+         */
+        const auto block_pass_or_net = [](auto event) {
```

Comment:
can we pull parts of this function into a static function (in the ShadowFSM struct) and add some unit tests?

Reply:
Added